include/awkward/array/NumpyArray.h

```diff
@@ -182,7 +183,11 @@ namespace awkward {
     /// The format may be specified explicitly.
     NumpyArray(const Index64 index, const std::string& format);
 
-    /// @brief Reference-counted pointer to the array buffer.
+    /// @brief A GPU Flag decides the location of the instance of this class
+    uint8_t memory_loc = -1; // This will be resolved once the content markers are in place
```

Comment:
This can't be a unsigned if `-1` is a possible value. Why not an `int64_t`? That's a small difference on our scale, and it's conceivable that a supercomputer would have more than 127 GPUs. (If not now, then someday.)

Anyway, this would be a good type to hide behind a new type name, probably in the `util::` namespace. Then, whether we change its meaning without changing its name or even if we do need to change its name or its surroundings (i.e. primitive types should be pass by value but more heavyweight class instances should be pass by reference, which means adding `const` and `&` to each argument), it would be easier to find all instances cleanly. Without a special name, we'd have to check all integers to find `memory_loc` variables.

Reply:
My bad, it was a hastily pushed PR. I'll change it to int64_t right away.

Yes, using a namespace makes sense. Is `util::` namespace good? Could it be something on the lines of `gpu::` since it's `gpu` exclusive? And, would it be a better idea to send these `memory_loc` via a reference. This is because, during the course of an operation, there might be a slight chance of a `Device to Device` Transfer when we push for Multi-GPU Programming. I'd have to read up on this, but in my understanding, this is not an impossible situation. It would be a good practice to update the `memory_loc` integer via a reference in those cases.

Reply:
First, on the namespace: even normal arrays in main memory would have a `memory_loc`; it's `-1`. The package without GPU support should, for simplicity, keep all the `memory_loc` logic in, even though it would always come to the conclusion that each buffer is in main memory. Trying to make that conditionally compile would be complex without much benefit.

Maybe this should live in the `kernel::` namespace, with the kernel function wrappers.

Thinking more carefully about this, perhaps `memory_loc` shouldn't be a property of array nodes (`Content` subclasses), but `Index`, `Identities`, `NumpyArray`, and `RawArray`, the four classes that directly hold buffers that can be in main memory or a GPU. Also, we can put off `Identities` and `RawArray` for later. Since the tree formed by `Content` subclasses is distributed, it can be hard to ensure that all of the nodes in this tree have the same `memory_loc`: a user can [ak.zip](https://awkward-array.readthedocs.io/en/latest/_auto/ak.zip.html) together some CPU arrays and some GPU arrays—when and _should_ they be copied over so that everything is on the same device? Maybe never: there's nothing in principle wrong with having an array in which some nodes are on a CPU and some nodes are on a GPU. After all, [VirtualArrays](https://awkward-array.readthedocs.io/en/latest/ak.layout.VirtualArray.html) allow some nodes to not even exist.

It might be more manageable if we don't try to control or even keep track of whether a `Content` subclass like `ListArray64` is "on the CPU" or "on the GPU" because its `starts` (an `Index`) might be on the CPU, its `stops` (another `Index`) might be on GPU 0 and its `content` (a `NumpyArray`, for example) might be on GPU 1. The only time when everything needs to be on the same device, and when we need to know what device that is, is when a kernel runs. All of the arrays involved in a single kernel need to be on the same device.

So that argues for tracking `memory_loc` on the `Index` class and on the `NumpyArray` class (both have `ptr`), ignoring `Identities` and `RawArray` for now. If we have to keep track of this, rather than CUDA, the `Index` and `NumpyArray` classes get a new member datum, `memory_loc`, and everywhere a kernel is called, the calling function has to pick one `memory_loc` and do call some `ensure_loc(preferred_memory_loc)` on all the `Indexes` involved in the kernel before the kernel.

But we might be able to do less at the calling sites and more in the kernel wrappers. Here's a research question: **Can CUDA determine the memory location of a raw pointer?** That is, given a pointer representing data in main memory and a pointer representing data in GPU 1, is there a CUDA function call that can tell us that? If so, _then we don't need to track the memory location at all_, we can just pass around pointers as we have been because the information is in the pointers themselves (and one of CUDA's internal lookup tables). Arguments to the wrapper functions would not include a `memory_loc`, but the wrapper functions would query each of the arrays in its arguments to find out which device they're on, pick one as a place to do the calculation, and copy all the ones that aren't there to be at that place.

If CUDA can identify the memory location of a raw pointer, then letting it do so would be much less effort and much less error prone than doing it ourselves, so please look into that. It would simplify the way the kernels are called because the calling function wouldn't need to decide which device to move everything to. However, calling sites are currently creating the output arrays and passing a pointer to be filled; if we take this simpler route in terms of not having to keep track of where everything is, the code will need to be changed to allow the kernel wrappers to create and return outputs. Instead of passing in an `Index`'s pointer to be filled, it would have to pass in an `Index` object reference to be replaced and a size to allocate, so that the kernel wrapper can allocate the new `Index` on the correct device, since only the kernel wrapper would have that information, not the calling site. That means more changes, but the prospect of tracking `memory_loc` manually is worse.

Regarding your question about device-to-device transfers: that can't happen automatically, can it? If it does, then we absolutely can't track it ourselves. Letting a `memory_loc` reference change in-place would break thread-safety. But if device-to-device transfers happen automatically, then CUDA must allow us to not specify where the array is, since we don't control it. Depending on how CUDA works, this `memory_loc` might not be a device-number-or-main-memory, it might just be a two-valued enum: CPU and GPU.

We need to know more about how CUDA works to make these design decisions.

Reply:
Discussed in the Gitter chats. 
`TODO:` Test out different `CUDA` features, for tracking and consistent CPU `<->` GPU transfers.